### PR TITLE
[WIP] Consistency models distillation examples

### PR DESCRIPTION
**Pull request by ayushtues**
# Add distillation training scripts for Consistency models

Continuation of https://github.com/dg845/diffusers/pull/2, adding distillation examples for the recently added Consistency Models ([paper](https://arxiv.org/abs/2303.01469), [original code](https://github.com/openai/consistency_models), [diffusers pipeline](https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/consistency_models))

## Model/Pipeline Description
Consistency Models ([paper](https://arxiv.org/abs/2303.01469), [code](https://github.com/openai/consistency_models)) are a new family of generative models similar to continuous-time diffusion models which support fast one-step generation. Diffusion models can be distilled into a consistency model for faster sampling, and consistency models can also be trained from scratch. From the paper abstract:

> Diffusion models have made significant breakthroughs in image, audio, and video generation, but they depend on an iterative generation process that causes slow sampling speed and caps their potential for real-time applications. To overcome this limitation, we propose consistency models, a new family of generative models that achieve high sample quality without adversarial training. They support fast one-step generation by design, while still allowing for few-step sampling to trade compute for sample quality. They also support zero-shot data editing, like image inpainting, colorization, and super-resolution, without requiring explicit training on these tasks. Consistency models can be trained either as a way to distill pre-trained diffusion models, or as standalone generative models. Through extensive experiments, we demonstrate that they outperform existing distillation techniques for diffusion models in one- and few-step generation. For example, we achieve the new state-of-the-art FID of 3.55 on CIFAR-10 and 6.20 on ImageNet 64x64 for one-step generation. When trained as standalone generative models, consistency models also outperform single-step, non-adversarial generative models on standard benchmarks like CIFAR-10, ImageNet 64x64 and LSUN 256x256.

In this PR, we implement the distillation procedure to distill a diffusion model into a consistency model as described in the paper

## TODO

- [ ] Add training scripts for distillation

CC
@dg845 
@williamberman 
@patrickvonplaten

### CODE REVIEW

**Comment by sayakpaul** [comment]
@ayushtues, thanks for this work! Let us know when you'd like a review from our end. The PR doesn't have to be super tight for an initial review just as an FYI.

**Comment by ayushtues** [comment]
Hey @sayakpaul yes I think it is good enough for an initial review right now, one major thing we need to figure out is how to do the scaling of the inputs for different teacher models (as mentioned in https://github.com/dg845/diffusers/pull/2#discussion_r1233164339). The original paper used EDM as a teacher model, and that was trained using a Karras Scheduler, so it followed the same scaling and processing as the Conistency model, but if we want to use something else like Stable Diffusion or DDPM as the teacher model, need to figure out what input scaling and processing to use for them

**Comment by ayushtues** [comment]
@dg845, @patrickvonplaten @williamberman @sayakpaul  do you think we should add examples of other applications of Consistency models mentioned in the paper like inpainting, colorization, super-resolution? They mostly seem to involve some clever processing of the inputs and the outputs of the model

**Comment by sayakpaul** [comment]
> one major thing we need to figure out is how to do the scaling of the inputs for different teacher models (as mentioned in [dg845#2 (comment)](https://github.com/dg845/diffusers/pull/2#discussion_r1233164339)).

If we decided to say, distill https://huggingface.co/valhalla/sd-pokemon-model, can't we use the same scheduler for distilling the student (with appropriate changes as needed) and use the same preprocessing steps? 

> @dg845, @patrickvonplaten @williamberman @sayakpaul do you think we should add examples of other applications of Consistency models mentioned in the paper like inpainting, colorization, super-resolution?

Let's gauge the community interest for that first and then we can definitely consider.

File: examples/consistency_models/script.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Comment (sayakpaul) [inline examples/consistency_models/script.sh:1]: We don't include shell-scripts in the examples.

File: examples/consistency_models/train_consistency_distillation.py [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+check_min_version("0.17.0.dev0")
+
+logger = get_logger(__name__, log_level="INFO")
+
+def append_dims(x, target_dims):
+    """Appends dimensions to the end of a tensor until it has target_dims dimensions."""
+    dims_to_append = target_dims - x.ndim
+    if dims_to_append < 0:
+        raise ValueError(f"input has {x.ndim} dims but target_dims is {target_dims}, which is less")
+    return x[(...,) + (None,) * dims_to_append]
```
Comment (sayakpaul) [inline examples/consistency_models/train_consistency_distillation.py:41]: Would prefer making these changes inline since these are small. 

Delegating these to a separate function kind of breaks the readability.

File: examples/consistency_models/train_consistency_distillation.py [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+    :param broadcast_shape: a larger shape of K dimensions with the batch
+                            dimension equal to the length of timesteps.
+    :return: a tensor of shape [batch_size, 1, ...] where the shape has K dims.
+    """
+    if not isinstance(arr, torch.Tensor):
+        arr = torch.from_numpy(arr)
+    res = arr[timesteps].float().to(timesteps.device)
+    while len(res.shape) < len(broadcast_shape):
+        res = res[..., None]
+    return res.expand(broadcast_shape)
```
Comment (sayakpaul) [inline examples/consistency_models/train_consistency_distillation.py:63]: Same.

File: examples/consistency_models/train_consistency_distillation.py [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+                    "ResnetUpsampleBlock2D",
+                ],
+                resnet_time_scale_shift="scale_shift",
+                upsample_type="resnet",
+                downsample_type="resnet"
+                )
+    else:
+        config = UNet2DModel.load_config(args.model_config_name_or_path)
+        model = UNet2DModel.from_config(config)
+        target_model = UNet2DModel.from_config(config)
```
Comment (sayakpaul) [inline examples/consistency_models/train_consistency_distillation.py:411]: So, this, in a sense, is step distillation i.e., we're not essentially compressing the model but reducing the number of steps.

File: examples/consistency_models/train_consistency_distillation.py [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+
+    
+    # load the model to distill into a consistency model
+    teacher_model = DDPMPipeline.from_pretrained("google/ddpm-cifar10-32").unet
+    model = model.float()
+    target_model = target_model.float() # TODO : support half precision training
+    teacher_model = teacher_model.float()
```
Comment (sayakpaul) [inline examples/consistency_models/train_consistency_distillation.py:418]: We don't need to do this explicitly when using `accelerator.prepare()`.

File: examples/consistency_models/train_consistency_distillation.py [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+                downsample_type="resnet"
+                )
+    else:
+        config = UNet2DModel.load_config(args.model_config_name_or_path)
+        model = UNet2DModel.from_config(config)
+        target_model = UNet2DModel.from_config(config)
+
+    
+    # load the model to distill into a consistency model
+    teacher_model = DDPMPipeline.from_pretrained("google/ddpm-cifar10-32").unet
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/consistency_models/train_consistency_distillation.py:415]: And it should also work with `DiffusionPipeline`.

Comment (sayakpaul) [inline examples/consistency_models/train_consistency_distillation.py:415]: Not clear what `model`, `target_model`, and `teacher_model` mean.

File: examples/consistency_models/train_consistency_distillation.py [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+        model, optimizer, noise_scheduler, train_dataloader, lr_scheduler, teacher_model, target_model, target_model_ema
+    )
+    noise_scheduler.set_timesteps(num_scales, device=accelerator.device)
+
+    target_model_ema.to(accelerator.device)
+    # We need to initialize the trackers we use, and also store our configuration.
+    # The trackers initializes automatically on the main process.
+    if accelerator.is_main_process:
+        run = os.path.split(__file__)[-1].split(".")[0]
+        accelerator.init_trackers(run)
```
Comment (sayakpaul) [inline examples/consistency_models/train_consistency_distillation.py:501]: We usually perform this initialization like so:

https://github.com/huggingface/diffusers/blob/7a91ea6c2b53f94da930a61ed571364022b21044/examples/dreambooth/train_dreambooth_lora.py#L1068

File: examples/consistency_models/train_consistency_distillation.py [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+            sigma = sigmas[index+1]
+            timestep_prev = timesteps[index]
+            sigma_prev = sigmas[index]
+            # add noise expects the scaled timestep only and internally converts to sigma
+            noised_image = noise_scheduler.add_noise(clean_images, noise, timestep)
+            target_model_ema.copy_to(target_model.parameters())
+
+            with accelerator.accumulate(model):
+                # Predict the noise residual
+                model_output = model(noise_scheduler.scale_model_input(noised_image, timestep), timestep, class_labels=labels).sample
```
Comment (sayakpaul) [inline examples/consistency_models/train_consistency_distillation.py:579]: There are three `model_output` variables here. Let's maybe try to give them more appropriate names.

**Review by sayakpaul — COMMENTED** [review]
Looking good. 

From my end, I think we need to:

- [ ] Have `accelerate` deeply and better integrated. 
- [ ] Improve the readability of the code by using better variable names, adding comments wherever applicable, and refactoring some bits wherever applicable. 
- [ ] Use [this](https://github.com/huggingface/diffusers/blob/7a91ea6c2b53f94da930a61ed571364022b21044/examples/dreambooth/train_dreambooth_lora.py#L1402) method to upload the trained checkpoints to the Hub. 
- [ ] Add tests and docs. 
- [ ] Add utility for performing validation inference runs with Weights and Biases. 
- [ ] Demonstrate a full run with reasonable results. 

More than happy to help with any of these :) 

Also, IIUC, Consistency Models can either be trained in isolation or can be used to step-distill a pre-trained model. If that's the case, should we allow both from the training script? Or am I wrong?

File: examples/consistency_models/train_consistency_distillation.py (hunk 6 above)
Comment (dg845) [reply examples/consistency_models/train_consistency_distillation.py:415]: My understanding is as follows (things in square brackets [] are the variable names in the current code):

- The `teacher_model` is the already-trained diffusion model we want to distill to a consistency model (for example, `google/ddpm-cifar10-32`).
- The `model` is a consistency model which is used to calculate the current output of the distilled consistency model $`\boldsymbol{f_\theta}(x_{t_{n + 1}}, t_{n + 1})`$ [`model_output` no. 1*] on the noisy sample $`x_{t_{n + 1}}`$ [`noised_image`] (following Algorithm 2 in the [paper](https://arxiv.org/pdf/2303.01469.pdf)). The parameters $`\boldsymbol{\theta}`$ are updated via gradient descent. (The paper calls this the "online" model, following RL terminology.)
- The `target_model` is a consistency model which is used to calculate the distillation target, which we want the consistency model output $`\boldsymbol{f_\theta}(x_{t_{n + 1}}, t_{n + 1})`$ to match. This is obtained by first running one step of the sampler (in the paper, one step of the Karras Heun sampler) on the `teacher_model` to obtain a previous noisy sample $`\hat{x}_{t_n}^\phi`$ [`model_output` no. 2*], and then getting the `target_model` consistency model's output $`\boldsymbol{f_{\theta^-}}(\hat{x}_{t_{n}}^\phi, t_{n})`$ [`model_output` no. 3*] on $`\hat{x}_{t_n}^\phi`$. The target parameters $`\boldsymbol{\theta^-}`$ are updated via an EMA update (using $`\boldsymbol{\theta}`$). (This is analogous to target models in [DQN-style](https://spinningup.openai.com/en/latest/algorithms/ddpg.html) RL algorithms, and I believe the EMA update is equivalent to what is sometimes called polyak averaging in RL.)

The loss function $`L^{\phi}(\boldsymbol{\theta}, \boldsymbol{\theta^-})`$ is then calculated using a distance metric $`d(\cdot, \cdot)`$ (MSE in the code):

```math
L^{\phi}(\boldsymbol{\theta}, \boldsymbol{\theta^-}) = d(\boldsymbol{f_\theta}(x_{t_{n + 1}}, t_{n + 1}), \boldsymbol{f_{\theta^-}}(\hat{x}_{t_{n}}^\phi, t_{n}))
```
and then we take a gradient descent step with respect to `model`'s parameters $`\boldsymbol{\theta}`$. (It's not clear to me which of `model` and `target_model` is outputted as the final trained checkpoint in the end; according to the paper, theoretically the parameters for the models should converge during training [e.g. $`\boldsymbol{\theta} = \boldsymbol{\theta^-}`$ at the end of training], so using either one should be fine.)

Some proposals:

- I agree that what `model`, `target_model`, `teacher_model` are doing might be hard to follow from the code alone, perhaps it might make sense to add comments explaining what each one does when they are initialized?
- Rename `model_output` no. 2 to `teacher_model_output` and `model_output` no. 3 to `target_model_output`. (If we want to rename `model` [see below], perhaps `model_output` no. 1 can be renamed similarly.)
- Perhaps renaming `model` to something like `online_model` or `current_consistency_model` might be more clear? (Not sure on this one, especially the `current_consistency_model` suggestion, since I'm not sure if it's accurate; would feel more comfortable with this if `model` is what is outputted as the final trained checkpoint, see above.)

(*) Technically, these quantities correspond to `distiller`, `denoised_image`, and `distiller_target` in the code, respectively, but they're all preceded by a calculated raw U-Net output named `model_output`.

File: examples/consistency_models/train_consistency_distillation.py (hunk 8 above)
Comment (dg845) [reply examples/consistency_models/train_consistency_distillation.py:579]: See https://github.com/huggingface/diffusers/pull/3992#discussion_r1259066402 for some suggestions.

File: examples/consistency_models/train_consistency_distillation.py (hunk 4 above)
Comment (dg845) [reply examples/consistency_models/train_consistency_distillation.py:411]: Perhaps it is a distinction without a difference, but I think of it as reducing the number of steps for a (diffusion model, sampler) pair. In my mind, this distinction is important because my guess is that the quality of the sampler affects the quality of the distilled consistency model; if we distill using a sampler which produces bad samples for the diffusion model, the consistency model might learn to faithfully produce similar samples to sampling from the diffusion model using the sampler, but which are still poor quality because the samples it learned from are poor quality. (Conversely, using a better sampler might produce a better consistency model.)

File: examples/consistency_models/train_consistency_distillation.py (hunk 6 above)
Comment (sayakpaul) [reply examples/consistency_models/train_consistency_distillation.py:415]: Thanks so much for explaining @dg845. See my responses below:

> I agree that what model, target_model, teacher_model are doing might be hard to follow from the code alone, perhaps it might make sense to add comments explaining what each one does when they are initialized?

100% agree. The math is heavy enough and we should grab the opportunity here to make it easily digestible to the end users. 

> Rename model_output no. 2 to teacher_model_output and model_output no. 3 to target_model_output. 

Works for me. 

> Perhaps renaming model to something like online_model or current_consistency_model might be more clear? (Not sure on this one, especially the current_consistency_model suggestion, since I'm not sure if it's accurate; would feel more comfortable with this if model is what is outputted as the final trained checkpoint, see above.)

Let's maybe try to output EMA'd model only since that is more common in practice?


> This is obtained by first running one step of the sampler (in the paper and code, one step of the Karras Heun sampler) on the teacher_model to obtain a previous noisy sample 
 [model_output no. 2*]

Have we checked if the `teacher_model` performs okay with that sampler? Does it matter?

**Comment by ayushtues** [comment]
Hey @sayakpaul can you help with the better integration with `accelerator`? I don't have much idea about how to use it better. Also can you link me to some relevant tests in existing examples which I can refer to

**Comment by sayakpaul** [comment]
> Hey @sayakpaul can you help with the better integration with accelerator? I don't have much idea about how to use it better.

If you follow one of our examples from here, I think you will have a better idea [this](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth.py) for example). 

> Also can you link me to some relevant tests in existing examples which I can refer to

Here you go: https://github.com/huggingface/diffusers/blob/main/examples/test_examples.py

File: examples/consistency_models/train_consistency_distillation.py (hunk 6 above)
Comment (sayakpaul) [reply examples/consistency_models/train_consistency_distillation.py:377]: Sounds like a plan! Thanks so much for being generous with your explanations!

File: src/diffusers/schedulers/scheduling_consistency_models.py [hunk 9]
```diff
@@ -268,6 +268,7 @@ def step(
         sample: torch.FloatTensor,
         generator: Optional[torch.Generator] = None,
         return_dict: bool = True,
+        use_noise: bool = True
```
[the remark this answers is not in the document]
Comment (williamberman) [reply src/diffusers/schedulers/scheduling_consistency_models.py:271]: I think it would be better to put the section of the function that computes the unnoised prediction in its own function that can be called. 

i.e.

```
def step(...):
    denoised = self.denoise(...)

    if len(self.timesteps) > 1:
         prev_sample = ...
    else:
        prev_sample = denoised

    return prev_sample

def denoise(...):
    ...
```

this would avoid extra global config to the class and the training script could just call denoise()

File: examples/consistency_models/train_consistency_distillation.py [hunk 10]
```diff
@@ -0,0 +1,770 @@
...
+                    "AttnDownBlock2D",
+                ],
+                up_block_types= [
+                    "AttnUpBlock2D",
+                    "ResnetUpsampleBlock2D",
+                ],
+                resnet_time_scale_shift="scale_shift",
+                upsample_type="resnet",
+                downsample_type="resnet"
+                )
```
Comment (williamberman) [inline examples/consistency_models/train_consistency_distillation.py:415]: could we remove this? would prefer to not hardcode a config in the training script

File: examples/consistency_models/train_consistency_distillation.py [hunk 11]
```diff
@@ -0,0 +1,770 @@
...
+                repo_name = args.hub_model_id
+            repo_id = create_repo(repo_name, exist_ok=True, token=args.hub_token).repo_id
+            repo = Repository(args.output_dir, clone_from=repo_name, token=args.hub_token)
+
+        elif args.output_dir is not None:
+            os.makedirs(args.output_dir, exist_ok=True)
+
+    # For testing use a dummy model
+    if args.testing: 
+        config = UNet2DModel.load_config('diffusers/consistency-models-test', subfolder="test_unet")
```
Comment (williamberman) [inline examples/consistency_models/train_consistency_distillation.py:367]: can we remove this? we should be able to pass the preferred model in from the standard cli args

File: src/diffusers/schedulers/scheduling_consistency_models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_consistency_models.py:271]: That would also work for me - is probably indeed the better option (as my first proposal would force people to update the config after training, which is probs more brittle)
